### PR TITLE
TDL-19644 Add date window for visitor stream

### PR DESCRIPTION
**Pull request by prijendev**
# Description of change
 - Added date window logic for the `visitors` stream.
 - Added unit test case for the code change.

# Manual QA steps
 -
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tests/unittests/test_date_window_for_get_records.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+                {'id': 1, 'metadata': {'auto': {"lastupdated": (current_time - timedelta(days=40)).timestamp() * 1000}}},
+                {'id': 2, 'metadata': {'auto': {"lastupdated": (current_time - timedelta(days=40)).timestamp() * 1000}}}
+            ],
+            [   # Second page of test-data with 10 days ago replication-key
+                {'id': 3, 'metadata': {'auto': {"lastupdated": (current_time - timedelta(days=10)).timestamp() * 1000}}}
+                
+            ]
+        ]
+
+        # Configure STream object and call get_records() function
```
Comment (dbshah1212) [inline tests/unittests/test_date_window_for_get_records.py:31]:
```suggestion
        # Configure stream object and call get_records() function
```

File: tap_pendo/streams.py [hunk 2]
```diff
@@ -1022,6 +1027,84 @@ def get_body(self):
...
+                max_bookmark = max(replication_value, max_bookmark)
+                yield record
+
+            # If data found in current date_window page then update_bookmark after yielding page of data.
+            if record:
+                self.update_bookmark(state, self.name, strftime(max_bookmark), self.replication_key)
+
+            # Set start_date and end_date of date window for next API call
+            start_dttm = end_dttm
+            end_dttm = start_dttm + timedelta(days=date_window_size)
```
Comment (dbshah1212) [inline tap_pendo/streams.py:1104]:
```suggestion
            end_dttm = end_dttm + timedelta(days=date_window_size)
```

File: tests/unittests/test_lazy_aggregation_sync.py [hunk 3]
```diff
@@ -65,11 +72,17 @@ def test_lazzy_aggregation_without_sub_stream(self, mocked_substream, mocked_sel
             Verify that if sub stream is not selected then also all data should be return for super stream
             and sync_substream should not be called 
         '''
-        expected_data = [{"id":1}, {"id":2}, {"id":3}]
-        records = '{"results": [{"id":1}, {"id":2}, {"id":3}]}'
+        expected_data = [
```
Comment (dbshah1212) [inline tests/unittests/test_lazy_aggregation_sync.py:75]: Add comment

File: tests/unittests/test_date_window_for_get_records.py (hunk 1 above)
Comment (prijendev) [reply tests/unittests/test_date_window_for_get_records.py]: Updated.

File: tap_pendo/streams.py (hunk 2 above)
Comment (prijendev) [reply tap_pendo/streams.py]: Replaced `start_dttm` with `end_dttm`.

**Review by savan-chovatiya — APPROVED** [review]

File: tap_pendo/streams.py [hunk 4]
```diff
@@ -1022,6 +1027,87 @@ def get_body(self):
             }
         }
 
+    def sync(self, state, start_date=None, key_id=None):
```
Comment (dbshah1212) [inline tap_pendo/streams.py:1030]: Add function comment

File: tap_pendo/streams.py [hunk 5]
```diff
@@ -997,9 +999,10 @@ class Visitors(LazyAggregationStream):
     name = "visitors"
     replication_method = "INCREMENTAL"
     replication_key = "lastupdated"
+    replication_key_path = "metadata.auto.lastupdated"
     key_properties = ["visitor_id"]
 
-    def get_body(self):
+    def get_body(self, start, end):
```
[the remark this answers is not in the document]
Comment (prijendev) [reply tap_pendo/streams.py:1006]: Added comment to explain start and end.

**Review by manand31 — APPROVED** [review]

**Pull request closed without merging**